src/rust/engine/options/src/args.rs

```diff
@@ -161,14 +161,21 @@ impl ArgsTracker {
         self.unconsumed_args.lock().remove(arg);
     }
 
-    pub fn get_unconsumed_flags(&self) -> Vec<String> {
-        let mut ret: Vec<String> = self
-            .unconsumed_args
-            .lock()
-            .iter()
-            .map(|arg| arg.flag.clone())
-            .collect();
-        ret.sort(); // For stability in tests and when reporting unconsumed args.
+    pub fn get_unconsumed_flags(&self) -> HashMap<Scope, Vec<String>> {
+        // Map from positional context (GLOBAL or a goal name) to unconsumed flags encountered
+        // at that position in the CLI args.
+        let mut ret: HashMap<Scope, Vec<String>> = HashMap::new();
+        for arg in self.unconsumed_args.lock().iter() {
+            if let Some(flags_for_context) = ret.get_mut(&arg.context) {
+                flags_for_context.push(arg.flag.clone());
+            } else {
+                let flags_for_context = vec![arg.flag.clone()];
+                ret.insert(arg.context.clone(), flags_for_context);
+            };
```

Comment:
FYI (no need to change, just flagging rust conventions/APIs), I think this sort of manipulation benefits a lot from the `entry` API, a bit like `defaultdict` (or `dict.setdefault()`) in Python: https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry

```suggestion
            ret.entry(args.context.clone()).or_insert(vec![]).push(arg.flag.clone());
```

(`or_insert_with` is handy too, if the default value is expensive (empty vecs are very cheap). This does force an extra clone of the `args.context` in the case that the key already exists which is unfortunate, but only does 1 (instead of 2) hashes in the case that the key doesn't exist... in any case, performance likely doesn't matter here.)

Reply:
Yeah, I did not do this because I didn't like the superfluous allocation. Admittedly this is meaningless in this specific case, but it makes the `entry` API a bit of an antipattern, so my bias is against using entry unless it improves performance, rather than being for it unless it harms performance.